### PR TITLE
625 new resource netapp ontap protocols fpolicy engine

### LINKED ISSUE #625: [New Resource]: netapp-ontap_protocols_fpolicy_engine
### Description

I am requesting a new resource to manage FPolicy engines in NetApp ONTAP. Currently, there is no Terraform resource to create and manage FPolicy engines.

FPolicy engines define the external servers that receive and process file operation notifications, including connection details, authentication methods, and communication protocols.

### Requested Resource(s) and/or Data Source(s)

* `netapp-ontap_protocols_fpolicy_external_engine` - Resource to create and manage FPolicy engines

### Potential Terraform Configuration

```terraform
# Create an FPolicy engine
resource "netapp-ontap_protocols_fpolicy_external_engine" "compliance_engine" {
  cx_profile_name = "cluster4"
  svm_name        = "carchi-test"
  name            = "compliance-engine"
  primary_servers = [
    "10.10.10.100",
    "10.10.10.101"
  ]
  port = 6789
  ssl_option = "no-auth"

  # High priority fields

  # --extern-engine-type
  type = "asynchronous"

  # -send-buffer-size
  send_buffer_size = 1048576

  # The above fields seem to be the ones we set in our old automation, there are ~15 fields for external engine
}
```

### References

- [NetApp ONTAP REST API - FPolicy Engines](https://docs.netapp.com/us-en/ontap-restmap/fpolicy.html#fpolicy-policy-external-engine-create)
- [FPolicy External Engine Create](https://docs.netapp.com/us-en/ontap-cli/vserver-fpolicy-policy-external-engine-create.html#parameters)

### Would you like to implement a fix?

None

### PR DESCRIPTION
**Pull request by DhivyaBharathi215**
1) Created New resource netapp-ontap_protocols_fpolicy_external_engine
2) Created New data sources netapp-ontap_protocols_fpolicy_external_engine and netapp-ontap_protocols_fpolicy_external_engines
3) ACC tests added
4) Updated docs for data sources and resources.

### CODE REVIEW

**Comment by chuyich** [comment]
In the examples, can you add the needed links like provider.tf,...

**Review by chuyich — CHANGES_REQUESTED** [review]
Missing the needed files under examples

**Review by suhasbshekar — APPROVED** [review]

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go [hunk 1]
```diff
@@ -0,0 +1,911 @@
...
+						Computed:            true,
+					},
+					"recv_buffer": schema.Int64Attribute{
+						MarkdownDescription: "Receive buffer size",
+						Optional:            true,
+						Computed:            true,
+					},
+				},
+			},
+			"secondary_servers": schema.ListAttribute{
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engine_resource.go]: if the sequence/order does not matter, it's better to use set. Ref. https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/provider/protocols/protocols_fpolicy_event_resource.go#L95

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go [hunk 2]
```diff
@@ -0,0 +1,911 @@
...
+				MarkdownDescription: "Specifies the ISO-8601 interval time for a storage appliance to query a status request from an FPolicy server",
+				Optional:            true,
+				Computed:            true,
+			},
+			"ssl_option": schema.StringAttribute{
+				MarkdownDescription: "The SSL option for external communication with the FPolicy server",
+				Optional:            true,
+				Computed:            true,
+			},
+			"primary_servers": schema.ListAttribute{
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engine_resource.go]: The order of elements in the list does not matter. It's better to use schema,SetAttribute. ref. https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/provider/protocols/protocols_fpolicy_event_resource.go#L95

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go [hunk 3]
```diff
@@ -0,0 +1,911 @@
...
+			},
+			"format": schema.StringAttribute{
+				MarkdownDescription: "The format for the notification messages sent to the FPolicy servers",
+				Optional:            true,
+				Computed:            true,
+			},
+			"type": schema.StringAttribute{
+				MarkdownDescription: "Determines what ONTAP does after sending notifications to FPolicy servers",
+				Optional:            true,
+				Computed:            true,
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engine_resource.go:345]: Since it's computed and optional, the followings need to be added:
`
				PlanModifiers: []planmodifier.String{
					stringplanmodifier.UseStateForUnknown(),
				},
`

ref. https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/provider/protocols/protocols_fpolicy_policy_resource.go#L103

**Review by chuyich — CHANGES_REQUESTED** [review]
There are many parameters are optional and computed. But they don't have 
				PlanModifiers: []planmodifier.String{
					stringplanmodifier.UseStateForUnknown(),
				},

Have you every tried to do "terraform apply" multiple times without resource file changes? Did terraform apply tell if there is a change when you do the second time?

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go (hunk 1 above)
Comment (DhivyaBharathi215) [reply internal/provider/protocols/protocols_fpolicy_external_engine_resource.go]: Updated the code

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go (hunk 2 above)
Comment (DhivyaBharathi215) [reply internal/provider/protocols/protocols_fpolicy_external_engine_resource.go]: Updated the code

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go (hunk 3 above)
Comment (DhivyaBharathi215) [reply internal/provider/protocols/protocols_fpolicy_external_engine_resource.go:345]: Updated the code

**Comment by DhivyaBharathi215** [comment]
> There are many parameters are optional and computed. But they don't have PlanModifiers: []planmodifier.String{ stringplanmodifier.UseStateForUnknown(), },
> 
> Have you every tried to do "terraform apply" multiple times without resource file changes? Did terraform apply tell if there is a change when you do the second time?

Didn't face any issue during second run without plan modifiers, still changed the code as per your comments. This also works fine.

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go (hunk 2 above)
Comment (chuyich) [reply internal/provider/protocols/protocols_fpolicy_external_engine_resource.go]: It's not resolved yet.

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go [hunk 4]
```diff
@@ -0,0 +1,952 @@
...
+	Name                  types.String `tfsdk:"name"`
+	SVMName               types.String `tfsdk:"svm_name"`
+	KeepAliveInterval     types.String `tfsdk:"keep_alive_interval"`
+	RequestCancelTimeout  types.String `tfsdk:"request_cancel_timeout"`
+	Certificate           types.Object `tfsdk:"certificate"`
+	SessionTimeout        types.String `tfsdk:"session_timeout"`
+	RequestAbortTimeout   types.String `tfsdk:"request_abort_timeout"`
+	StatusRequestInterval types.String `tfsdk:"status_request_interval"`
+	SSLOption             types.String `tfsdk:"ssl_option"`
+	PrimaryServers        types.List   `tfsdk:"primary_servers"`
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engine_resource.go]: ref https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/provider/protocols/protocols_s3_group_resource.go#L51

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go [hunk 5]
```diff
@@ -0,0 +1,952 @@
...
+	BufferSize            types.Object `tfsdk:"buffer_size"`
+	SecondaryServers      types.List   `tfsdk:"secondary_servers"`
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engine_resource.go]: ref. https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/provider/protocols/protocols_s3_group_resource.go#L51

File: internal/provider/protocols/protocols_fpolicy_external_engines_data_source.go [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+						},
+						"status_request_interval": schema.StringAttribute{
+							MarkdownDescription: "Specifies the ISO-8601 interval time for a storage appliance to query a status request from an FPolicy server",
+							Computed:            true,
+						},
+						"ssl_option": schema.StringAttribute{
+							MarkdownDescription: "The SSL option for external communication with the FPolicy server",
+							Computed:            true,
+						},
+					"primary_servers": schema.SetAttribute{
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engines_data_source.go]: Indent

File: internal/provider/protocols/protocols_fpolicy_external_engines_data_source.go [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+									MarkdownDescription: "Send buffer size",
+									Computed:            true,
+								},
+								"recv_buffer": schema.Int64Attribute{
+									MarkdownDescription: "Receive buffer size",
+									Computed:            true,
+								},
+							},
+						},
+					"secondary_servers": schema.SetAttribute{
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engines_data_source.go]: indent

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go [hunk 8]
```diff
@@ -0,0 +1,952 @@
...
+	resiliencyObjectValue, diags := types.ObjectValue(resiliencyElementTypes, resiliencyElements)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+	data.Resiliency = resiliencyObjectValue
+
+	// Set primary servers
+	if len(restInfo.PrimaryServers) > 0 {
+		data.PrimaryServers, _ = types.ListValueFrom(ctx, types.StringType, restInfo.PrimaryServers)
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engine_resource.go]: ref .https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/provider/protocols/protocols_s3_group_resource.go#L185

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go [hunk 9]
```diff
@@ -0,0 +1,952 @@
...
+	}
+
+	// Set secondary servers
+	if len(restInfo.SecondaryServers) > 0 {
+		data.SecondaryServers, _ = types.ListValueFrom(ctx, types.StringType, restInfo.SecondaryServers)
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engine_resource.go]: ref. https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/provider/protocols/protocols_s3_group_resource.go#L185

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go [hunk 10]
```diff
@@ -0,0 +1,952 @@
...
+		// Special handling for enabled=false - ensure directory_path is provided
+		// ONTAP requires directory_path when any resiliency config is sent
+		if !resiliency.Enabled.ValueBool() && resiliency.DirectoryPath.IsNull() {
+			// Set a default path when disabling resiliency
+			body.Resiliency.DirectoryPath = ""
+		}
+	}
+
+	// Set primary servers
+	if !data.PrimaryServers.IsNull() {
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engine_resource.go:759]: ref. https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/provider/protocols/protocols_s3_group_resource.go#L247

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go [hunk 11]
```diff
@@ -0,0 +1,952 @@
...
+		var primaryServers []string
+		resp.Diagnostics.Append(data.PrimaryServers.ElementsAs(ctx, &primaryServers, false)...)
+		body.PrimaryServers = primaryServers
+	}
+
+	// Set secondary servers
+	if !data.SecondaryServers.IsNull() {
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engine_resource.go:766]: Same. ref . https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/provider/protocols/protocols_s3_group_resource.go#L247

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go [hunk 12]
```diff
@@ -0,0 +1,952 @@
...
+	}
+	resiliencyObjectValue, diagsResiliency := types.ObjectValue(resiliencyElementTypes, resiliencyElements)
+	if diagsResiliency.HasError() {
+		resp.Diagnostics.Append(diagsResiliency...)
+		return
+	}
+	data.Resiliency = resiliencyObjectValue
+
+	// Set primary servers
+	if restInfo.PrimaryServers != nil {
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engine_resource.go:922]: ref. https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/provider/protocols/protocols_s3_group_resource.go#L421

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go [hunk 13]
```diff
@@ -0,0 +1,952 @@
...
+		primaryServersList, diagsPrimary := types.ListValueFrom(ctx, types.StringType, restInfo.PrimaryServers)
+		if diagsPrimary.HasError() {
+			resp.Diagnostics.Append(diagsPrimary...)
+			return
+		}
+		data.PrimaryServers = primaryServersList
+	}
+
+	// Set secondary servers
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engine_resource.go:935]: ref. https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/provider/protocols/protocols_s3_group_resource.go#L421

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go [hunk 14]
```diff
@@ -0,0 +1,952 @@
...
+	}
+	resiliencyObjectValue, diagsResiliency := types.ObjectValue(resiliencyElementTypes, resiliencyElements)
+	if diagsResiliency.HasError() {
+		resp.Diagnostics.Append(diagsResiliency...)
+		return
+	}
+	data.Resiliency = resiliencyObjectValue
+
+	// Set primary servers
+	if restInfo.PrimaryServers != nil {
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engine_resource.go:1254]: https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/provider/protocols/protocols_s3_group_resource.go#L421

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go [hunk 15]
```diff
@@ -0,0 +1,952 @@
...
+		primaryServersList, diagsPrimary := types.ListValueFrom(ctx, types.StringType, restInfo.PrimaryServers)
+		if diagsPrimary.HasError() {
+			resp.Diagnostics.Append(diagsPrimary...)
+			return
+		}
+		data.PrimaryServers = primaryServersList
+	}
+
+	// Set secondary servers
+	if restInfo.SecondaryServers != nil {
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engine_resource.go:1268]: ref. https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/provider/protocols/protocols_s3_group_resource.go#L421

File: internal/provider/protocols/protocols_fpolicy_external_engines_data_source.go [hunk 16]
```diff
@@ -0,0 +1,392 @@
...
+						Required:            true,
+					},
+					"name": schema.StringAttribute{
+						MarkdownDescription: "FPolicy external engine name",
+						Optional:            true,
+					},
+				},
+				Optional: true,
+			},
+			"protocols_fpolicy_external_engines": schema.ListNestedAttribute{
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engines_data_source.go:71]: Should be set not list?

**Review by chuyich — CHANGES_REQUESTED** [review]
List vs Set: please update accordingly

File: internal/provider/protocols/protocols_fpolicy_external_engines_data_source.go (hunk 16 above)
Comment (DhivyaBharathi215) [reply internal/provider/protocols/protocols_fpolicy_external_engines_data_source.go:71]: Verified with other modules, all data source modules are defined as List attribute .

**Comment by DhivyaBharathi215** [comment]
> List vs Set: please update accordingly



> List vs Set: please update accordingly


Somehow my resource changes for primary and secondary servers were not pushed properly,
now updated the code as per the comments.

File: internal/provider/protocols/protocols_fpolicy_external_engine_data_source.go [hunk 17]
```diff
@@ -0,0 +1,393 @@
...
+				Computed:            true,
+			},
+			"primary_servers": schema.SetAttribute{
+				MarkdownDescription: "The primary FPolicy servers to which the node sends",
+				Computed:            true,
+				ElementType:         types.StringType,
+			},
+			"buffer_size": schema.SingleNestedAttribute{
+				MarkdownDescription: "Specifies the send and receive buffer size of the connected socket for the FPolicy server",
+				Optional:            true,
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engine_data_source.go]: I think it's not optional. It should be computed only.

File: internal/provider/protocols/protocols_fpolicy_external_engine_data_source.go [hunk 18]
```diff
@@ -0,0 +1,393 @@
...
+		})
+	}
+
+	// Set primary servers
+	if len(restInfo.PrimaryServers) > 0 {
+		primaryServersList := make([]types.String, len(restInfo.PrimaryServers))
+		for i, server := range restInfo.PrimaryServers {
+			primaryServersList[i] = types.StringValue(server)
+		}
+		data.PrimaryServers, _ = types.SetValue(types.StringType, basetypes.ListValue{}.Elements())
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engine_data_source.go]: Line 363 and Line 365 are kind of repeated. Not sure which one is what you want. Line 365 should be sufficent.

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go [hunk 19]
```diff
@@ -0,0 +1,1014 @@
...
+		"recv_buffer": types.Int64Value(int64(restInfo.BufferSize.RecvBuffer)),
+	}
+	bufferObjectValue, diags := types.ObjectValue(bufferElementTypes, bufferElements)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+	data.BufferSize = bufferObjectValue
+
+	// Set resiliency using ObjectValue
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engine_resource.go]: certificate/buffer_size/resiliency are always set using types.ObjectValue(...) with whatever came back from ONTAP, even if fields are empty strings/zeros. That means state will always contain an object (not null), which can cause:

perpetual diffs when config omits these blocks (null vs empty object),
or inability to distinguish “unset” vs “set to empty”.

Can you test these cases (combination like no resilience, empty resilience) and see if only create it with required parameters. And terraform apply again. See if it looks fine.

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go [hunk 20]
```diff
@@ -0,0 +1,1014 @@
...
+			DirectoryPath:     resiliency.DirectoryPath.ValueString(),
+			RetentionDuration: resiliency.RetentionDuration.ValueString(),
+			Enabled:           resiliency.Enabled.ValueBool(),
+		}
+		
+		// Special handling for enabled=false - ensure directory_path is provided
+		// ONTAP requires directory_path when any resiliency config is sent
+		if !resiliency.Enabled.ValueBool() && resiliency.DirectoryPath.IsNull() {
+			// Set a default path when disabling resiliency
+			body.Resiliency.DirectoryPath = ""
```
Comment (chuyich) [inline internal/provider/protocols/protocols_fpolicy_external_engine_resource.go]: if enable is false, directory_path is provided, then why it needs to be set ""? 
If requires directory_path whenever resiliency is sent, enforce it with a schema validator (or at least a diag error) when resiliency block is present, regardless of enabled true/false.
If ONTAP only requires it when enabled=true, then only require it in that case and don’t send the resiliency block at all when enabled=false and other fields are null/empty.
A little bit confused.

**Review by chuyich — CHANGES_REQUESTED** [review]
Can you test some edge cases to make sure the obj with empty or not declare work properly with multiple terraform apply does not show the diff.

File: internal/provider/protocols/protocols_fpolicy_external_engine_data_source.go (hunk 17 above)
Comment (DhivyaBharathi215) [reply internal/provider/protocols/protocols_fpolicy_external_engine_data_source.go]: If buffer size is not configured by user then ONTAP returns default value.

File: internal/provider/protocols/protocols_fpolicy_external_engine_data_source.go (hunk 18 above)
Comment (DhivyaBharathi215) [reply internal/provider/protocols/protocols_fpolicy_external_engine_data_source.go]: Removed

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go (hunk 19 above)
Comment (DhivyaBharathi215) [reply internal/provider/protocols/protocols_fpolicy_external_engine_resource.go]: Now handled the NULL values and if the cert value is not passed instead of setting "" it will be null.

File: internal/provider/protocols/protocols_fpolicy_external_engine_resource.go (hunk 20 above)
Comment (DhivyaBharathi215) [reply internal/provider/protocols/protocols_fpolicy_external_engine_resource.go]: If the values are not passed then getting,
 REST reported error restclient.RestError{Code:"9765007", Message:"A required
│ parameter is missing for configuring the resiliency feature. Specify
│ \"-is-resiliency-enabled true\" with \" -resiliency-directory-path\" and,
│ optionally, \"-resiliency-max-retention-duration\".",

**Comment by DhivyaBharathi215** [comment]
> Can you test some edge cases to make sure the obj with empty or not declare work properly with multiple terraform apply does not show the diff.

resource "netapp-ontap_protocols_fpolicy_external_engine" "default" {
  cx_profile_name = "cluster4"
  svm_name        = "vs0"
  name            = "minimal_test"
  type            = "synchronous"
  port            = 9876
  format          = "xml"
  primary_servers = ["10.0.1.100"] # no certificate, buffer_size, or resiliency blocks # should result in null values in state, not empty objects
}

resource "netapp-ontap_protocols_fpolicy_external_engine" "cert_name_only" {
  cx_profile_name = "cluster4"
  svm_name        = "vs0"
  name            = "cert_name_only_test"
  type            = "synchronous"
  port            = 9878
  format          = "xml"
  primary_servers = ["10.0.1.102"]  
  certificate = {
    name = "fpolicy_cert" # ca and serial_number omitted
  }

resource "netapp-ontap_protocols_fpolicy_external_engine" "full_config" {
  cx_profile_name = "cluster4"
  svm_name        = "vs0"
  name            = "full_config_test"
  type            = "asynchronous"
  port            = 9888
  format          = "xml"
  primary_servers = ["10.0.1.112", "10.0.1.113"]
  secondary_servers = ["10.0.1.114", "10.0.1.115"]
  certificate = {
    name          = "fpolicy_ssl_cert"
    ca            = "internal_root_ca"
    serial_number = "1234567890ABCDEF"
  }
  buffer_size = {
    send_buffer = 8192
    recv_buffer = 16384
  }  
  resiliency = {
    enabled            = true
    directory_path     = "/"
    retention_duration = "PT2M"
  }
  keep_alive_interval      = "PT50S"
  request_cancel_timeout   = "PT20S"  
  session_timeout          = "PT20S"
  request_abort_timeout    = "PT40S"
  status_request_interval  = "PT40S"
  server_progress_timeout  = "PT1M40S" PT100S
  ssl_option              = "no_auth"
}

resource "netapp-ontap_protocols_fpolicy_external_engine" "cert_only" { # only certificate configured
  cx_profile_name = "cluster4"
  svm_name        = "vs0"
  name            = "cert_only_test"
  type            = "synchronous"
  port            = 9889
  format          = "xml"
  primary_servers = ["10.0.1.116"]
  certificate = {
    name = "ssl_cert"
    ca   = "company_ca"
  } # buffer_size and resiliency omitted - should be null
}

resource "netapp-ontap_protocols_fpolicy_external_engine" "buffer_only" {
  cx_profile_name = "cluster4"
  svm_name        = "vs0"
  name            = "buffer_only_test"
  type            = "synchronous"
  port            = 9890
  format          = "xml"
  primary_servers = ["10.0.1.117"]
  buffer_size = {  # Only buffer_size configured
    send_buffer = 4096
    recv_buffer = 8192
  } # certificate and resiliency omitted - should be null
}

resource "netapp-ontap_protocols_fpolicy_external_engine" "resiliency_only" { # Only resiliency configured
  cx_profile_name = "cluster4"
  svm_name        = "vs0"
  name            = "resiliency_only_test"
  type            = "asynchronous"
  port            = 9891
  format          = "xml"
  primary_servers = ["10.0.1.118"]
  resiliency = {
    enabled        = true
    directory_path = "/"
  } # certificate and buffer_size omitted - should be null
}

File: internal/provider/protocols/protocols_fpolicy_external_engine_data_source.go (hunk 17 above)
Comment (chuyich) [reply internal/provider/protocols/protocols_fpolicy_external_engine_data_source.go]: Since it's a data source, so the value will be from GET. Not from the user input, right? The another data source also treat this is computed only.

File: internal/interfaces/protocols_fpolicy_external_engine.go [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+// ProtocolsFpolicyExternalEngineCertificate describes the certificate data model
+type ProtocolsFpolicyExternalEngineCertificate struct {
+	SerialNumber string `mapstructure:"serial_number,omitempty"`
+	Name         string `mapstructure:"name,omitempty"`
+	Ca           string `mapstructure:"ca,omitempty"`
+}
+
+// ProtocolsFpolicyExternalEngineBufferSize describes the buffer size data model
+type ProtocolsFpolicyExternalEngineBufferSize struct {
+	SendBuffer int `mapstructure:"send_buffer,omitempty"`
```
Comment (chuyich) [inline internal/interfaces/protocols_fpolicy_external_engine.go]: both type should be int64 (for consistency)

File: docs/data-sources/fpolicy_external_engine.md [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+
+# Related ONTAP commands
+
+fpolicy policy external-engine show
+
+## Supported Platforms
+
+- On-perm ONTAP system 9.6 or higher
+- Amazon FSx for NetApp ONTAP
+- The parameters `max_server_requests`, `resiliency`, `format`, `server_progress_timeout`, `buffer_size`, `ssl_option`, `status_request_interval`, `status_request_interval`, `certificate` and `request_cancel_timeout` can only be used in ONTAP 9.11 or higher
```
Comment (chuyich) [inline docs/data-sources/fpolicy_external_engine.md]: Two 'status_request_interval'

File: docs/data-sources/fpolicy_external_engines.md [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+
+# Related ONTAP commands
+
+fpolicy policy external-engine show
+
+## Supported Platforms
+
+- On-perm ONTAP system 9.6 or higher
+- Amazon FSx for NetApp ONTAP
+- The parameters `max_server_requests`, `resiliency`, `format`, `server_progress_timeout`, `buffer_size`, `ssl_option`, `status_request_interval`, `status_request_interval`, `certificate` and `request_cancel_timeout` can only be used in ONTAP 9.11 or higher
```
Comment (chuyich) [inline docs/data-sources/fpolicy_external_engines.md]: two 'status_request_interval'

File: docs/resources/fpolicy_external_engine.md [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+* fpolicy policy external-engine create
+* fpolicy policy external-engine modify
+* fpolicy policy external-engine delete
+```
+
+## Supported Platforms
+
+- On-perm ONTAP system 9.6 or higher
+- Amazon FSx for NetApp ONTAP
+- The parameters `max_server_requests`, `resiliency`, `format`, `server_progress_timeout`, `buffer_size`, `ssl_option`, `status_request_interval`, `status_request_interval`, `certificate` and `request_cancel_timeout` can only be used in ONTAP 9.11 or higher
```
Comment (chuyich) [inline docs/resources/fpolicy_external_engine.md]: two 'status_request_interval' - repeated

File: docs/data-sources/fpolicy_external_engine.md [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+
+Retrieves FPolicy engine configuration data source
+
+# Related ONTAP commands
+
+fpolicy policy external-engine show
+
+## Supported Platforms
+
+- On-perm ONTAP system 9.6 or higher
```
Comment (chuyich) [inline docs/data-sources/fpolicy_external_engine.md]: typo: On-prem

File: docs/data-sources/fpolicy_external_engines.md [hunk 26]
```diff
@@ -0,0 +1,115 @@
...
+
+Retrieves FPolicy engines configuration data source
+
+# Related ONTAP commands
+
+fpolicy policy external-engine show
+
+## Supported Platforms
+
+- On-perm ONTAP system 9.6 or higher
```
Comment (chuyich) [inline docs/data-sources/fpolicy_external_engines.md]: typo: On-prem

File: docs/resources/fpolicy_external_engine.md [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+
+```commandline
+* fpolicy policy external-engine create
+* fpolicy policy external-engine modify
+* fpolicy policy external-engine delete
+```
+
+## Supported Platforms
+
+- On-perm ONTAP system 9.6 or higher
```
Comment (chuyich) [inline docs/resources/fpolicy_external_engine.md]: typo: on-prem

**Review by chuyich — CHANGES_REQUESTED** [review]
There are some places need to be fixed. Also, would you please try the copilot (on the top right of the PR) to do the code review and see anything were missed.

File: internal/interfaces/protocols_fpolicy_external_engine.go (hunk 21 above)
Comment (DhivyaBharathi215) [reply internal/interfaces/protocols_fpolicy_external_engine.go]: Resolved

File: docs/data-sources/fpolicy_external_engine.md (hunk 22 above)
Comment (DhivyaBharathi215) [reply docs/data-sources/fpolicy_external_engine.md]: Resolved

File: docs/data-sources/fpolicy_external_engines.md (hunk 23 above)
Comment (DhivyaBharathi215) [reply docs/data-sources/fpolicy_external_engines.md]: Resolved

File: docs/resources/fpolicy_external_engine.md (hunk 24 above)
Comment (DhivyaBharathi215) [reply docs/resources/fpolicy_external_engine.md]: Resolved

File: docs/data-sources/fpolicy_external_engine.md (hunk 25 above)
Comment (DhivyaBharathi215) [reply docs/data-sources/fpolicy_external_engine.md]: Resolved

File: docs/data-sources/fpolicy_external_engines.md (hunk 26 above)
Comment (DhivyaBharathi215) [reply docs/data-sources/fpolicy_external_engines.md]: Resolved

File: docs/resources/fpolicy_external_engine.md (hunk 27 above)
Comment (DhivyaBharathi215) [reply docs/resources/fpolicy_external_engine.md]: Resolved

**Comment by DhivyaBharathi215** [comment]
Fixed almost all valid suggestions from Github copilot, and also updated few logics
1. Normalized duration format (PT90S vs PT1M30S) for consistency.
2. Renamed data source and resource as per standard.

The module is working as expected. tested the following scenarios
# Test Scenario 1: Basic FPolicy external engine with all configuration options for testing modifications
resource "netapp-ontap_fpolicy_external_engine" "basic" {
  cx_profile_name = "cluster6"
  name           = "basic_engine"
  svm_name       = "tf_acc_svm"
  port           = 9002
  primary_servers = [
    "10.132.145.20"
  ]

# Test Scenario 2: FPolicy external engine with certificate configuration
resource "netapp-ontap_fpolicy_external_engine" "with_certificate" {
  cx_profile_name = "cluster6"
  name           = "cert_engine"
  svm_name       = "tf_acc_svm"
  port           = 9003
  primary_servers = [
    "10.132.145.21"
  ]
  ssl_option = "server_auth"
  certificate = {
    name          = "fpolicy_cert"
    serial_number = "1234567890"
    ca            = "fpolicy_ca"
  }
}

# Test Scenario 3: FPolicy external engine with buffer size configuration
resource "netapp-ontap_fpolicy_external_engine" "with_buffer" {
  cx_profile_name = "cluster6"
  name           = "buffer_engine"
  svm_name       = "tf_acc_svm"
  port           = 9004
  primary_servers = [
    "10.132.145.22"
  ]
  buffer_size = {
    send_buffer = 1024000
    recv_buffer = 2048000
  }
}

# Test Scenario 4: FPolicy external engine with resiliency enabled
resource "netapp-ontap_fpolicy_external_engine" "with_resiliency" {
  cx_profile_name = "cluster6"
  name           = "resilient_engine"
  svm_name       = "tf_acc_svm"
  port           = 9005
  primary_servers = [
    "10.132.145.23"
  ]
  type = "asynchronous"
  resiliency = {
    enabled           = true
    directory_path    = "/"          # Use root path which should exist
    retention_duration = "PT600S"    # Max allowed: 10 minutes
  }
}

# Test Scenario 5: FPolicy external engine with secondary servers
resource "netapp-ontap_fpolicy_external_engine" "with_secondary" {
  cx_profile_name = "cluster6"
  name           = "secondary_engine"
  svm_name       = "tf_acc_svm"
  port           = 9006
  primary_servers = [
    "10.132.145.24"
  ]
  secondary_servers = [
    "10.132.145.25",
    "10.132.145.26"
  ]
}

# Test Scenario 6: Duration normalization testing
resource "netapp-ontap_fpolicy_external_engine" "duration_test" {
  cx_profile_name = "cluster6"
  name           = "duration_engine"
  svm_name       = "tf_acc_svm"
  port           = 9007
  primary_servers = [
    "10.132.145.27"
  ] # Test various duration formats that should be normalized by ONTAP
  keep_alive_interval     = "PT180S"    # Should normalize to PT3M
  session_timeout         = "PT90S"     # Should normalize to PT1M30S
  request_cancel_timeout  = "PT100S"    # Max allowed value
  request_abort_timeout   = "PT200S"    # Max allowed value
  status_request_interval = "PT50S"     # Max allowed value
  server_progress_timeout = "PT100S"    # Max allowed value (was PT600S)
}

# Test Scenario 7: Full-featured FPolicy external engine
resource "netapp-ontap_fpolicy_external_engine" "protocols_fpolicy_external_engine" {
  cx_profile_name = "cluster6"
  name           = "full_featured_test"
  svm_name       = "tf_acc_svm"
  port           = 9008
  primary_servers = [
    "10.132.145.28",
    "10.132.145.29"
  ]

  secondary_servers = [
    "10.132.145.30"
  ]

  type = "asynchronous"
  format = "protobuf"

  keep_alive_interval     = "PT3M"
  session_timeout         = "PT15S"
  request_cancel_timeout  = "PT30S"
  request_abort_timeout   = "PT1M"
  status_request_interval = "PT15S"
  server_progress_timeout = "PT1M21S"

  max_server_requests    = 860
  max_connection_retries = 10

  ssl_option = "no_auth"
  certificate = { # Certificate configuration (when using SSL)
    name          = "fpolicy_cert1"
    serial_number = "12345678922"
    ca            = "fpolicy_test_ca"
  }

  buffer_size = {
    send_buffer = 1424000
    recv_buffer = 4494304
  }

  resiliency = {
    enabled            = true
    directory_path     = "/"
    retention_duration = "PT2M"
  }
}

# Test Scenario 8: XML format with synchronous type
resource "netapp-ontap_fpolicy_external_engine" "xml_synchronous" {
  cx_profile_name = "cluster6"
  name           = "xml_sync_engine"
  svm_name       = "tf_acc_svm"
  port           = 9009
  primary_servers = [
    "10.132.145.31"
  ]
  type   = "synchronous"
  format = "xml"
  ssl_option = "no_auth"
}

**Review by chuyich — APPROVED** [review]
Thanks for showing the edge cases!
